{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example of how to use the new design program            \n",
    "dec 2018        \n",
    "Gabriel Besombes  \n",
    "##    \n",
    "#### We will use:\n",
    "* reference file : \"/output/genomic/plant/Actinidia/chinensis/CK51F3_01/Genome/Assembly/PS1/1.68.5/AllChromosomes/PS1.1.68.5.fasta\"\n",
    "* annotation file : \"/output/genomic/plant/Actinidia/chinensis/Resequencing/Variants/PS1.1.68.5/52DiploidGenomes/Combined_diploidCK_basic_NS30_Q50_SAFR3_DP50_PAIR0.8_PS1.1.68.5_ann.vcf.gz\"\n",
    "#     \n",
    "#        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports:\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import python file from other directory in Jupyter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run \"../design.py\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "#    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating a PrimerDesign object:\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Without a target file and with default parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = PrimerDesign(reference_file=\"/output/genomic/plant/Actinidia/chinensis/CK51F3_01/Genome/Assembly/PS1/1.68.5/AllChromosomes/PS1.1.68.5.fasta\",\n",
    "                 annotation_file=\"/output/genomic/plant/Actinidia/chinensis/Resequencing/Variants/PS1.1.68.5/52DiploidGenomes/Combined_diploidCK_basic_NS30_Q50_SAFR3_DP50_PAIR0.8_PS1.1.68.5_ann.vcf.gz\",\n",
    "                 description=\"PS1.1.68.5\",\n",
    "                 targets_file=None,\n",
    "                 output_dir=None,\n",
    "                 amplicon_size_range=None,\n",
    "                 primer_size_range=None,\n",
    "                 p3_globals=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "#      "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quick test design on a small region:\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With auto you can do a quick design with default parameters.   \n",
    "It will :\n",
    "* Create targets in the specifyed region if no target file was given for the PrimerDesign object and output these in the output directory in a bed file called \"targets.bed\"\n",
    "* Create a csv and a bed for the amplicons, same for the primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.auto(region={\"CHR1\":[[0, 2500], [5000, 7500]], \"CHR2\":[[0, 10000]]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>PRIMER_INTERNAL_0</th>\n",
       "      <th>PRIMER_INTERNAL_0_GC_PERCENT</th>\n",
       "      <th>PRIMER_INTERNAL_0_HAIRPIN_TH</th>\n",
       "      <th>PRIMER_INTERNAL_0_PENALTY</th>\n",
       "      <th>PRIMER_INTERNAL_0_SELF_ANY_TH</th>\n",
       "      <th>PRIMER_INTERNAL_0_SELF_END_TH</th>\n",
       "      <th>PRIMER_INTERNAL_0_SEQUENCE</th>\n",
       "      <th>PRIMER_INTERNAL_0_TM</th>\n",
       "      <th>PRIMER_INTERNAL_1</th>\n",
       "      <th>...</th>\n",
       "      <th>PRIMER_RIGHT_1_GC_PERCENT</th>\n",
       "      <th>PRIMER_RIGHT_1_HAIRPIN_TH</th>\n",
       "      <th>PRIMER_RIGHT_1_PENALTY</th>\n",
       "      <th>PRIMER_RIGHT_1_SELF_ANY_TH</th>\n",
       "      <th>PRIMER_RIGHT_1_SELF_END_TH</th>\n",
       "      <th>PRIMER_RIGHT_1_SEQUENCE</th>\n",
       "      <th>PRIMER_RIGHT_1_TM</th>\n",
       "      <th>PRIMER_RIGHT_EXPLAIN</th>\n",
       "      <th>PRIMER_RIGHT_NUM_RETURNED</th>\n",
       "      <th>REF_OFFSET</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>(141, 22)</td>\n",
       "      <td>54.545455</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.660472</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCCAACGCCCACCATGAATGCA</td>\n",
       "      <td>59.339528</td>\n",
       "      <td>(141, 22)</td>\n",
       "      <td>...</td>\n",
       "      <td>47.619048</td>\n",
       "      <td>32.4404</td>\n",
       "      <td>2.800250</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AATTGACCCTGAAGCAGAAGG</td>\n",
       "      <td>58.199750</td>\n",
       "      <td>considered 317, overlap excluded region 35, lo...</td>\n",
       "      <td>2</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>considered 165, low tm 3, high tm 86, ok 76</td>\n",
       "      <td>0</td>\n",
       "      <td>166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>(152, 20)</td>\n",
       "      <td>70.000000</td>\n",
       "      <td>33.236722</td>\n",
       "      <td>0.230110</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CCGAGACTTGGCTGACCCGC</td>\n",
       "      <td>59.769890</td>\n",
       "      <td>(152, 20)</td>\n",
       "      <td>...</td>\n",
       "      <td>40.909091</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4.974645</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TTCCTATGAACAGACAACAGGT</td>\n",
       "      <td>57.025355</td>\n",
       "      <td>considered 507, GC content failed 85, low tm 3...</td>\n",
       "      <td>2</td>\n",
       "      <td>329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>(152, 20)</td>\n",
       "      <td>70.000000</td>\n",
       "      <td>33.236722</td>\n",
       "      <td>0.230110</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CCGAGACTTGGCTGACCCGC</td>\n",
       "      <td>59.769890</td>\n",
       "      <td>(152, 20)</td>\n",
       "      <td>...</td>\n",
       "      <td>40.909091</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4.974645</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TTCCTATGAACAGACAACAGGT</td>\n",
       "      <td>57.025355</td>\n",
       "      <td>considered 57, low tm 43, high hairpin stabili...</td>\n",
       "      <td>2</td>\n",
       "      <td>329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>considered 190, overlap excluded region 36, GC...</td>\n",
       "      <td>0</td>\n",
       "      <td>454</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROMOSOME PRIMER_INTERNAL_0  PRIMER_INTERNAL_0_GC_PERCENT  \\\n",
       "0       CHR1         (141, 22)                     54.545455   \n",
       "1       CHR1               NaN                           NaN   \n",
       "2       CHR1         (152, 20)                     70.000000   \n",
       "3       CHR1         (152, 20)                     70.000000   \n",
       "4       CHR1               NaN                           NaN   \n",
       "\n",
       "   PRIMER_INTERNAL_0_HAIRPIN_TH  PRIMER_INTERNAL_0_PENALTY  \\\n",
       "0                      0.000000                   2.660472   \n",
       "1                           NaN                        NaN   \n",
       "2                     33.236722                   0.230110   \n",
       "3                     33.236722                   0.230110   \n",
       "4                           NaN                        NaN   \n",
       "\n",
       "   PRIMER_INTERNAL_0_SELF_ANY_TH  PRIMER_INTERNAL_0_SELF_END_TH  \\\n",
       "0                            0.0                            0.0   \n",
       "1                            NaN                            NaN   \n",
       "2                            0.0                            0.0   \n",
       "3                            0.0                            0.0   \n",
       "4                            NaN                            NaN   \n",
       "\n",
       "  PRIMER_INTERNAL_0_SEQUENCE  PRIMER_INTERNAL_0_TM PRIMER_INTERNAL_1  \\\n",
       "0     TCCAACGCCCACCATGAATGCA             59.339528         (141, 22)   \n",
       "1                        NaN                   NaN               NaN   \n",
       "2       CCGAGACTTGGCTGACCCGC             59.769890         (152, 20)   \n",
       "3       CCGAGACTTGGCTGACCCGC             59.769890         (152, 20)   \n",
       "4                        NaN                   NaN               NaN   \n",
       "\n",
       "     ...      PRIMER_RIGHT_1_GC_PERCENT  PRIMER_RIGHT_1_HAIRPIN_TH  \\\n",
       "0    ...                      47.619048                    32.4404   \n",
       "1    ...                            NaN                        NaN   \n",
       "2    ...                      40.909091                     0.0000   \n",
       "3    ...                      40.909091                     0.0000   \n",
       "4    ...                            NaN                        NaN   \n",
       "\n",
       "   PRIMER_RIGHT_1_PENALTY  PRIMER_RIGHT_1_SELF_ANY_TH  \\\n",
       "0                2.800250                         0.0   \n",
       "1                     NaN                         NaN   \n",
       "2                4.974645                         0.0   \n",
       "3                4.974645                         0.0   \n",
       "4                     NaN                         NaN   \n",
       "\n",
       "   PRIMER_RIGHT_1_SELF_END_TH PRIMER_RIGHT_1_SEQUENCE  PRIMER_RIGHT_1_TM  \\\n",
       "0                         0.0   AATTGACCCTGAAGCAGAAGG          58.199750   \n",
       "1                         NaN                     NaN                NaN   \n",
       "2                         0.0  TTCCTATGAACAGACAACAGGT          57.025355   \n",
       "3                         0.0  TTCCTATGAACAGACAACAGGT          57.025355   \n",
       "4                         NaN                     NaN                NaN   \n",
       "\n",
       "                                PRIMER_RIGHT_EXPLAIN  \\\n",
       "0  considered 317, overlap excluded region 35, lo...   \n",
       "1        considered 165, low tm 3, high tm 86, ok 76   \n",
       "2  considered 507, GC content failed 85, low tm 3...   \n",
       "3  considered 57, low tm 43, high hairpin stabili...   \n",
       "4  considered 190, overlap excluded region 36, GC...   \n",
       "\n",
       "   PRIMER_RIGHT_NUM_RETURNED REF_OFFSET  \n",
       "0                          2        114  \n",
       "1                          0        166  \n",
       "2                          2        329  \n",
       "3                          2        329  \n",
       "4                          0        454  \n",
       "\n",
       "[5 rows x 70 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.p3_out.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "#     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bed files analysis:\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All of the bed files associated with the PrimerDesign object can be analysed using the data in the vcf anotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.amplicons.analyse(d.annotations._pyvcf_reader, d.reference, write=\"analysed.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will create a pandas dataframe that looks like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>SSRs</th>\n",
       "      <th>SSRs_variants</th>\n",
       "      <th>PI</th>\n",
       "      <th>min_call_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHR2</td>\n",
       "      <td>75</td>\n",
       "      <td>297</td>\n",
       "      <td>{'AAACCAA': [2], 'CCCTAAA': [3]}</td>\n",
       "      <td>{'CHR2:136-145': {'CCTAAGC': [2]}, 'CHR2:227-2...</td>\n",
       "      <td>4.801258</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHR2</td>\n",
       "      <td>74</td>\n",
       "      <td>297</td>\n",
       "      <td>{'AAACCAA': [2], 'CCCTAAA': [3]}</td>\n",
       "      <td>{'CHR2:136-145': {'CCTAAGC': [2]}, 'CHR2:227-2...</td>\n",
       "      <td>4.801258</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHR2</td>\n",
       "      <td>203</td>\n",
       "      <td>413</td>\n",
       "      <td>{'AT': [11], 'AGA': [6], 'GAA': [4]}</td>\n",
       "      <td>{'CHR2:227-236': {'AAACCTG': [3, 2]}}</td>\n",
       "      <td>1.364420</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHR2</td>\n",
       "      <td>203</td>\n",
       "      <td>412</td>\n",
       "      <td>{'AT': [11], 'AGA': [6], 'GAA': [4]}</td>\n",
       "      <td>{'CHR2:227-236': {'AAACCTG': [3, 2]}}</td>\n",
       "      <td>1.364420</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>159</td>\n",
       "      <td>414</td>\n",
       "      <td>{'CA': [30, 24, 21, 12], 'AT': [24], 'TAT': [4...</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.145193</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR START  END                                               SSRs  \\\n",
       "0  CHR2    75  297                   {'AAACCAA': [2], 'CCCTAAA': [3]}   \n",
       "1  CHR2    74  297                   {'AAACCAA': [2], 'CCCTAAA': [3]}   \n",
       "2  CHR2   203  413               {'AT': [11], 'AGA': [6], 'GAA': [4]}   \n",
       "3  CHR2   203  412               {'AT': [11], 'AGA': [6], 'GAA': [4]}   \n",
       "4  CHR1   159  414  {'CA': [30, 24, 21, 12], 'AT': [24], 'TAT': [4...   \n",
       "\n",
       "                                       SSRs_variants        PI  min_call_rate  \n",
       "0  {'CHR2:136-145': {'CCTAAGC': [2]}, 'CHR2:227-2...  4.801258            1.0  \n",
       "1  {'CHR2:136-145': {'CCTAAGC': [2]}, 'CHR2:227-2...  4.801258            1.0  \n",
       "2              {'CHR2:227-236': {'AAACCTG': [3, 2]}}  1.364420            1.0  \n",
       "3              {'CHR2:227-236': {'AAACCTG': [3, 2]}}  1.364420            1.0  \n",
       "4                                                 {}  0.145193            1.0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.amplicons.analysed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can then filter this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.amplicons.filter()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That will also create a pandas dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>SSRs</th>\n",
       "      <th>SSRs_variants</th>\n",
       "      <th>PI</th>\n",
       "      <th>min_call_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHR2</td>\n",
       "      <td>75</td>\n",
       "      <td>297</td>\n",
       "      <td>{'AAACCAA': [2], 'CCCTAAA': [3]}</td>\n",
       "      <td>{'CHR2:136-145': {'CCTAAGC': [2]}, 'CHR2:227-2...</td>\n",
       "      <td>4.801258</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHR2</td>\n",
       "      <td>74</td>\n",
       "      <td>297</td>\n",
       "      <td>{'AAACCAA': [2], 'CCCTAAA': [3]}</td>\n",
       "      <td>{'CHR2:136-145': {'CCTAAGC': [2]}, 'CHR2:227-2...</td>\n",
       "      <td>4.801258</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHR2</td>\n",
       "      <td>203</td>\n",
       "      <td>413</td>\n",
       "      <td>{'AT': [11], 'AGA': [6], 'GAA': [4]}</td>\n",
       "      <td>{'CHR2:227-236': {'AAACCTG': [3, 2]}}</td>\n",
       "      <td>1.364420</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHR2</td>\n",
       "      <td>203</td>\n",
       "      <td>412</td>\n",
       "      <td>{'AT': [11], 'AGA': [6], 'GAA': [4]}</td>\n",
       "      <td>{'CHR2:227-236': {'AAACCTG': [3, 2]}}</td>\n",
       "      <td>1.364420</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHR2</td>\n",
       "      <td>203</td>\n",
       "      <td>456</td>\n",
       "      <td>{'AT': [12], 'AGA': [8], 'GAA': [7], 'CCT': [2...</td>\n",
       "      <td>{'CHR2:227-236': {'AAACCTG': [3, 2]}}</td>\n",
       "      <td>1.401797</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR START  END                                               SSRs  \\\n",
       "0  CHR2    75  297                   {'AAACCAA': [2], 'CCCTAAA': [3]}   \n",
       "1  CHR2    74  297                   {'AAACCAA': [2], 'CCCTAAA': [3]}   \n",
       "2  CHR2   203  413               {'AT': [11], 'AGA': [6], 'GAA': [4]}   \n",
       "3  CHR2   203  412               {'AT': [11], 'AGA': [6], 'GAA': [4]}   \n",
       "4  CHR2   203  456  {'AT': [12], 'AGA': [8], 'GAA': [7], 'CCT': [2...   \n",
       "\n",
       "                                       SSRs_variants        PI  min_call_rate  \n",
       "0  {'CHR2:136-145': {'CCTAAGC': [2]}, 'CHR2:227-2...  4.801258            1.0  \n",
       "1  {'CHR2:136-145': {'CCTAAGC': [2]}, 'CHR2:227-2...  4.801258            1.0  \n",
       "2              {'CHR2:227-236': {'AAACCTG': [3, 2]}}  1.364420            1.0  \n",
       "3              {'CHR2:227-236': {'AAACCTG': [3, 2]}}  1.364420            1.0  \n",
       "4              {'CHR2:227-236': {'AAACCTG': [3, 2]}}  1.401797            1.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.amplicons.filtered.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can then create a new bed file from this filtered result and point the object to this new file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.amplicons.replace(name=\"filtered_targets.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'filtered_targets'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.amplicons.name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "#      "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## More detailed and flexible design:\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can specify a target file directly during the object creation or afterwards like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.targets = \"targets.bed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you don't have a targets file you can create it like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.gettargets(region={\"CHR1\":[[0, 2500], [5000, 7500]], \"CHR11\":[[0, 5000]]}, write=\"advanced_targets.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And change the targtes file used by the PrimerDesign object like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.targets = \"advanced_targets.bed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can then analyse and filter your targets on different criteria:\n",
    "* SSR contents of the region\n",
    "* SSR contents in the variants\n",
    "* Nucleotide diversity\n",
    "* Minimum call_rate in the region   \n",
    "     \n",
    "If any is set to False it will not be analysed.     \n",
    "These result will be available in a pandas dataframe and can be outputed to a csv format if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.targets.analyse(d.annotations._pyvcf_reader,\n",
    "                  d.reference,\n",
    "                  SSRs=True,\n",
    "                  SSRs_variants=True,\n",
    "                  PI=True,\n",
    "                  call_rate=True,\n",
    "                  write=\"analysed_targets.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>SSRs</th>\n",
       "      <th>SSRs_variants</th>\n",
       "      <th>PI</th>\n",
       "      <th>min_call_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>215</td>\n",
       "      <td>394</td>\n",
       "      <td>{'CA': [22, 19, 10], 'TAT': [4], 'TTG': [3], '...</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.145193</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>215</td>\n",
       "      <td>446</td>\n",
       "      <td>{'CA': [26, 23, 14], 'TAT': [5], 'TTG': [4], '...</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.695238</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>435</td>\n",
       "      <td>609</td>\n",
       "      <td>{'CACTC': [2], 'TC': [14], 'GA': [10], 'TCT': ...</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.016173</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>360</td>\n",
       "      <td>609</td>\n",
       "      <td>{'CT': [24], 'CACTC': [2], 'TC': [14], 'GA': [...</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.123989</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>532</td>\n",
       "      <td>734</td>\n",
       "      <td>{'TCT': [6], 'TGT': [3], 'TATT': [3], 'CTA': [3]}</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.503504</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR START  END                                               SSRs  \\\n",
       "0  CHR1   215  394  {'CA': [22, 19, 10], 'TAT': [4], 'TTG': [3], '...   \n",
       "1  CHR1   215  446  {'CA': [26, 23, 14], 'TAT': [5], 'TTG': [4], '...   \n",
       "2  CHR1   435  609  {'CACTC': [2], 'TC': [14], 'GA': [10], 'TCT': ...   \n",
       "3  CHR1   360  609  {'CT': [24], 'CACTC': [2], 'TC': [14], 'GA': [...   \n",
       "4  CHR1   532  734  {'TCT': [6], 'TGT': [3], 'TATT': [3], 'CTA': [3]}   \n",
       "\n",
       "  SSRs_variants        PI  min_call_rate  \n",
       "0            {}  0.145193            1.0  \n",
       "1            {}  0.695238            1.0  \n",
       "2            {}  1.016173            1.0  \n",
       "3            {}  1.123989            1.0  \n",
       "4            {}  0.503504            1.0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.targets.analysed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The formats and defaults for the filtering are as follow:\n",
    "* SSRs in the region : SSRs = {\n",
    "                2: 10,\n",
    "                3: 2,\n",
    "                4: 2,\n",
    "                5: 2,\n",
    "                6: 2,\n",
    "                7: 2,\n",
    "                8: 2,\n",
    "                9: 2,\n",
    "                10: 2\n",
    "            }\n",
    "* SSRs in the variants : SSRs_variants = {\n",
    "                2: 10,\n",
    "                3: 2,\n",
    "                4: 2,\n",
    "                5: 2,\n",
    "                6: 2,\n",
    "                7: 2,\n",
    "                8: 2,\n",
    "                9: 2,\n",
    "                10: 2\n",
    "            }\n",
    "* Nucleotide diversity : PI = 0.5\n",
    "* Minimum call_rate in the region : call_rate = 0.5\n",
    "\n",
    "If any is set to False it will not be used for filtering.   \n",
    "These result will be available in a pandas dataframe and can be outputed to a csv format if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.targets.filter(write=\"filtered_targets.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see that only the targets that had a PI>0.5, min_call_rate>0.5, SSRs and SSRs in variants of length 2 that repeated 10 or more times, or length 3 to 10 that repeated 2 or more times where kept."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>SSRs</th>\n",
       "      <th>SSRs_variants</th>\n",
       "      <th>PI</th>\n",
       "      <th>min_call_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5658</td>\n",
       "      <td>5861</td>\n",
       "      <td>{'GGGAGG': [2], 'AG': [12, 9], 'TCT': [5], 'GC...</td>\n",
       "      <td>{'CHR1:5843-5861': {'AGCG': [3]}}</td>\n",
       "      <td>0.838994</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5623</td>\n",
       "      <td>5861</td>\n",
       "      <td>{'GGGAGG': [2], 'AG': [12, 9], 'TCT': [5], 'GC...</td>\n",
       "      <td>{'CHR1:5843-5861': {'AGCG': [3]}}</td>\n",
       "      <td>0.838994</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5773</td>\n",
       "      <td>5934</td>\n",
       "      <td>{'CT': [6], 'CA': [15, 8], 'AT': [10, 7], 'AGC...</td>\n",
       "      <td>{'CHR1:5843-5861': {'AGCG': [3]}, 'CHR1:5914-5...</td>\n",
       "      <td>1.259659</td>\n",
       "      <td>0.962264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5734</td>\n",
       "      <td>5934</td>\n",
       "      <td>{'TCT': [5], 'GCT': [3], 'TG': [7], 'CT': [6],...</td>\n",
       "      <td>{'CHR1:5843-5861': {'AGCG': [3]}, 'CHR1:5914-5...</td>\n",
       "      <td>1.297035</td>\n",
       "      <td>0.962264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5821</td>\n",
       "      <td>5989</td>\n",
       "      <td>{'TA': [8], 'CA': [9], 'AT': [8], 'AGCG': [4],...</td>\n",
       "      <td>{'CHR1:5843-5861': {'AGCG': [3]}, 'CHR1:5914-5...</td>\n",
       "      <td>1.295597</td>\n",
       "      <td>0.962264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR START   END                                               SSRs  \\\n",
       "0  CHR1  5658  5861  {'GGGAGG': [2], 'AG': [12, 9], 'TCT': [5], 'GC...   \n",
       "1  CHR1  5623  5861  {'GGGAGG': [2], 'AG': [12, 9], 'TCT': [5], 'GC...   \n",
       "2  CHR1  5773  5934  {'CT': [6], 'CA': [15, 8], 'AT': [10, 7], 'AGC...   \n",
       "3  CHR1  5734  5934  {'TCT': [5], 'GCT': [3], 'TG': [7], 'CT': [6],...   \n",
       "4  CHR1  5821  5989  {'TA': [8], 'CA': [9], 'AT': [8], 'AGCG': [4],...   \n",
       "\n",
       "                                       SSRs_variants        PI  min_call_rate  \n",
       "0                  {'CHR1:5843-5861': {'AGCG': [3]}}  0.838994       1.000000  \n",
       "1                  {'CHR1:5843-5861': {'AGCG': [3]}}  0.838994       1.000000  \n",
       "2  {'CHR1:5843-5861': {'AGCG': [3]}, 'CHR1:5914-5...  1.259659       0.962264  \n",
       "3  {'CHR1:5843-5861': {'AGCG': [3]}, 'CHR1:5914-5...  1.297035       0.962264  \n",
       "4  {'CHR1:5843-5861': {'AGCG': [3]}, 'CHR1:5914-5...  1.295597       0.962264  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.targets.filtered.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can then change the targets file like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.targets.replace(\"filtered_targets.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can check that it worked:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'filtered_targets.bed'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.targets.file_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### You can now do the design on different region or without specifying a region:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.run_p3()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This only outputs crude results from primer3 to self.p3_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>PRIMER_INTERNAL_0</th>\n",
       "      <th>PRIMER_INTERNAL_0_GC_PERCENT</th>\n",
       "      <th>PRIMER_INTERNAL_0_HAIRPIN_TH</th>\n",
       "      <th>PRIMER_INTERNAL_0_PENALTY</th>\n",
       "      <th>PRIMER_INTERNAL_0_SELF_ANY_TH</th>\n",
       "      <th>PRIMER_INTERNAL_0_SELF_END_TH</th>\n",
       "      <th>PRIMER_INTERNAL_0_SEQUENCE</th>\n",
       "      <th>PRIMER_INTERNAL_0_TM</th>\n",
       "      <th>PRIMER_INTERNAL_1</th>\n",
       "      <th>...</th>\n",
       "      <th>PRIMER_RIGHT_1_GC_PERCENT</th>\n",
       "      <th>PRIMER_RIGHT_1_HAIRPIN_TH</th>\n",
       "      <th>PRIMER_RIGHT_1_PENALTY</th>\n",
       "      <th>PRIMER_RIGHT_1_SELF_ANY_TH</th>\n",
       "      <th>PRIMER_RIGHT_1_SELF_END_TH</th>\n",
       "      <th>PRIMER_RIGHT_1_SEQUENCE</th>\n",
       "      <th>PRIMER_RIGHT_1_TM</th>\n",
       "      <th>PRIMER_RIGHT_EXPLAIN</th>\n",
       "      <th>PRIMER_RIGHT_NUM_RETURNED</th>\n",
       "      <th>REF_OFFSET</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>(258, 20)</td>\n",
       "      <td>65.0</td>\n",
       "      <td>41.005307</td>\n",
       "      <td>0.113248</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>TTGACAGCGAGCGAGCGAGC</td>\n",
       "      <td>60.113248</td>\n",
       "      <td>(258, 20)</td>\n",
       "      <td>...</td>\n",
       "      <td>55.0</td>\n",
       "      <td>45.415575</td>\n",
       "      <td>0.677731</td>\n",
       "      <td>0.864186</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>TTCCAATGGAGCTTCTGCGG</td>\n",
       "      <td>60.677731</td>\n",
       "      <td>considered 133, overlap excluded region 46, lo...</td>\n",
       "      <td>2</td>\n",
       "      <td>5581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>(258, 20)</td>\n",
       "      <td>65.0</td>\n",
       "      <td>41.005307</td>\n",
       "      <td>0.113248</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>TTGACAGCGAGCGAGCGAGC</td>\n",
       "      <td>60.113248</td>\n",
       "      <td>(258, 20)</td>\n",
       "      <td>...</td>\n",
       "      <td>55.0</td>\n",
       "      <td>40.049105</td>\n",
       "      <td>0.455005</td>\n",
       "      <td>13.838267</td>\n",
       "      <td>5.482251</td>\n",
       "      <td>GCTTCCAATGGAGCTTCTGC</td>\n",
       "      <td>59.544995</td>\n",
       "      <td>considered 98, overlap excluded region 11, low...</td>\n",
       "      <td>2</td>\n",
       "      <td>5581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>(185, 20)</td>\n",
       "      <td>65.0</td>\n",
       "      <td>41.005307</td>\n",
       "      <td>0.113248</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>TTGACAGCGAGCGAGCGAGC</td>\n",
       "      <td>60.113248</td>\n",
       "      <td>(185, 20)</td>\n",
       "      <td>...</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.179029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ACACCAGTTACACCACCAGC</td>\n",
       "      <td>60.179029</td>\n",
       "      <td>considered 475, overlap excluded region 25, lo...</td>\n",
       "      <td>2</td>\n",
       "      <td>5654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>(185, 20)</td>\n",
       "      <td>65.0</td>\n",
       "      <td>41.005307</td>\n",
       "      <td>0.113248</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>TTGACAGCGAGCGAGCGAGC</td>\n",
       "      <td>60.113248</td>\n",
       "      <td>(185, 20)</td>\n",
       "      <td>...</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034896</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ACCACCCGTCAACTCACATC</td>\n",
       "      <td>59.965104</td>\n",
       "      <td>considered 241, overlap excluded region 25, lo...</td>\n",
       "      <td>2</td>\n",
       "      <td>5654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>(130, 20)</td>\n",
       "      <td>65.0</td>\n",
       "      <td>41.005307</td>\n",
       "      <td>0.113248</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>TTGACAGCGAGCGAGCGAGC</td>\n",
       "      <td>60.113248</td>\n",
       "      <td>(130, 20)</td>\n",
       "      <td>...</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.179032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CACCAGTTACACCACCAGCA</td>\n",
       "      <td>60.179032</td>\n",
       "      <td>considered 433, overlap excluded region 25, lo...</td>\n",
       "      <td>2</td>\n",
       "      <td>5709</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROMOSOME PRIMER_INTERNAL_0  PRIMER_INTERNAL_0_GC_PERCENT  \\\n",
       "0       CHR1         (258, 20)                          65.0   \n",
       "1       CHR1         (258, 20)                          65.0   \n",
       "2       CHR1         (185, 20)                          65.0   \n",
       "3       CHR1         (185, 20)                          65.0   \n",
       "4       CHR1         (130, 20)                          65.0   \n",
       "\n",
       "   PRIMER_INTERNAL_0_HAIRPIN_TH  PRIMER_INTERNAL_0_PENALTY  \\\n",
       "0                     41.005307                   0.113248   \n",
       "1                     41.005307                   0.113248   \n",
       "2                     41.005307                   0.113248   \n",
       "3                     41.005307                   0.113248   \n",
       "4                     41.005307                   0.113248   \n",
       "\n",
       "   PRIMER_INTERNAL_0_SELF_ANY_TH  PRIMER_INTERNAL_0_SELF_END_TH  \\\n",
       "0                       6.191775                       6.191775   \n",
       "1                       6.191775                       6.191775   \n",
       "2                       6.191775                       6.191775   \n",
       "3                       6.191775                       6.191775   \n",
       "4                       6.191775                       6.191775   \n",
       "\n",
       "  PRIMER_INTERNAL_0_SEQUENCE  PRIMER_INTERNAL_0_TM PRIMER_INTERNAL_1  \\\n",
       "0       TTGACAGCGAGCGAGCGAGC             60.113248         (258, 20)   \n",
       "1       TTGACAGCGAGCGAGCGAGC             60.113248         (258, 20)   \n",
       "2       TTGACAGCGAGCGAGCGAGC             60.113248         (185, 20)   \n",
       "3       TTGACAGCGAGCGAGCGAGC             60.113248         (185, 20)   \n",
       "4       TTGACAGCGAGCGAGCGAGC             60.113248         (130, 20)   \n",
       "\n",
       "     ...      PRIMER_RIGHT_1_GC_PERCENT  PRIMER_RIGHT_1_HAIRPIN_TH  \\\n",
       "0    ...                           55.0                  45.415575   \n",
       "1    ...                           55.0                  40.049105   \n",
       "2    ...                           55.0                   0.000000   \n",
       "3    ...                           55.0                   0.000000   \n",
       "4    ...                           55.0                   0.000000   \n",
       "\n",
       "   PRIMER_RIGHT_1_PENALTY  PRIMER_RIGHT_1_SELF_ANY_TH  \\\n",
       "0                0.677731                    0.864186   \n",
       "1                0.455005                   13.838267   \n",
       "2                0.179029                    0.000000   \n",
       "3                0.034896                    0.000000   \n",
       "4                0.179032                    0.000000   \n",
       "\n",
       "   PRIMER_RIGHT_1_SELF_END_TH PRIMER_RIGHT_1_SEQUENCE  PRIMER_RIGHT_1_TM  \\\n",
       "0                    0.000000    TTCCAATGGAGCTTCTGCGG          60.677731   \n",
       "1                    5.482251    GCTTCCAATGGAGCTTCTGC          59.544995   \n",
       "2                    0.000000    ACACCAGTTACACCACCAGC          60.179029   \n",
       "3                    0.000000    ACCACCCGTCAACTCACATC          59.965104   \n",
       "4                    0.000000    CACCAGTTACACCACCAGCA          60.179032   \n",
       "\n",
       "                                PRIMER_RIGHT_EXPLAIN  \\\n",
       "0  considered 133, overlap excluded region 46, lo...   \n",
       "1  considered 98, overlap excluded region 11, low...   \n",
       "2  considered 475, overlap excluded region 25, lo...   \n",
       "3  considered 241, overlap excluded region 25, lo...   \n",
       "4  considered 433, overlap excluded region 25, lo...   \n",
       "\n",
       "   PRIMER_RIGHT_NUM_RETURNED REF_OFFSET  \n",
       "0                          2       5581  \n",
       "1                          2       5581  \n",
       "2                          2       5654  \n",
       "3                          2       5654  \n",
       "4                          2       5709  \n",
       "\n",
       "[5 rows x 70 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.p3_out.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want a bit more than this you can use self.cleanoutput()."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.cleanoutput()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This creates pandas dataframe for the amplicons and the primers and export these as csv and bed files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>END</th>\n",
       "      <th>PRIMER_INTERNAL</th>\n",
       "      <th>PRIMER_INTERNAL_EXPLAIN</th>\n",
       "      <th>PRIMER_INTERNAL_GC_PERCENT</th>\n",
       "      <th>PRIMER_INTERNAL_HAIRPIN_TH</th>\n",
       "      <th>PRIMER_INTERNAL_NUM_RETURNED</th>\n",
       "      <th>PRIMER_INTERNAL_PENALTY</th>\n",
       "      <th>PRIMER_INTERNAL_REGION</th>\n",
       "      <th>PRIMER_INTERNAL_SELF_ANY_TH</th>\n",
       "      <th>...</th>\n",
       "      <th>PRIMER_RIGHT_HAIRPIN_TH</th>\n",
       "      <th>PRIMER_RIGHT_NUM_RETURNED</th>\n",
       "      <th>PRIMER_RIGHT_PENALTY</th>\n",
       "      <th>PRIMER_RIGHT_REGION</th>\n",
       "      <th>PRIMER_RIGHT_SELF_ANY_TH</th>\n",
       "      <th>PRIMER_RIGHT_SELF_END_TH</th>\n",
       "      <th>PRIMER_RIGHT_SEQUENCE</th>\n",
       "      <th>PRIMER_RIGHT_TM</th>\n",
       "      <th>REF_OFFSET</th>\n",
       "      <th>START</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CHR11</td>\n",
       "      <td>395</td>\n",
       "      <td>(195, 20)</td>\n",
       "      <td>considered 3185, GC content failed 3, low tm 2...</td>\n",
       "      <td>70.0</td>\n",
       "      <td>41.637802</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.033108</td>\n",
       "      <td>CHR11:287-307</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.799681</td>\n",
       "      <td>CHR11:372-395</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ACTCTACTACTCGAAGTTAGCAC</td>\n",
       "      <td>57.200319</td>\n",
       "      <td>92.0</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CHR11</td>\n",
       "      <td>395</td>\n",
       "      <td>(195, 20)</td>\n",
       "      <td>considered 3185, GC content failed 3, low tm 2...</td>\n",
       "      <td>70.0</td>\n",
       "      <td>41.637802</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.033108</td>\n",
       "      <td>CHR11:287-307</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.799681</td>\n",
       "      <td>CHR11:372-395</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ACTCTACTACTCGAAGTTAGCAC</td>\n",
       "      <td>57.200319</td>\n",
       "      <td>92.0</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5894</td>\n",
       "      <td>(258, 20)</td>\n",
       "      <td>considered 3005, GC content failed 5, low tm 2...</td>\n",
       "      <td>65.0</td>\n",
       "      <td>41.005307</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.113248</td>\n",
       "      <td>CHR1:5839-5859</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>...</td>\n",
       "      <td>40.049105</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.455005</td>\n",
       "      <td>CHR1:5874-5894</td>\n",
       "      <td>13.838267</td>\n",
       "      <td>5.482251</td>\n",
       "      <td>GCTTCCAATGGAGCTTCTGC</td>\n",
       "      <td>59.544995</td>\n",
       "      <td>5581.0</td>\n",
       "      <td>5599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5894</td>\n",
       "      <td>(258, 20)</td>\n",
       "      <td>considered 3355, GC content failed 5, low tm 2...</td>\n",
       "      <td>65.0</td>\n",
       "      <td>41.005307</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.113248</td>\n",
       "      <td>CHR1:5839-5859</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>...</td>\n",
       "      <td>40.049105</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.455005</td>\n",
       "      <td>CHR1:5874-5894</td>\n",
       "      <td>13.838267</td>\n",
       "      <td>5.482251</td>\n",
       "      <td>GCTTCCAATGGAGCTTCTGC</td>\n",
       "      <td>59.544995</td>\n",
       "      <td>5581.0</td>\n",
       "      <td>5634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5894</td>\n",
       "      <td>(258, 20)</td>\n",
       "      <td>considered 3005, GC content failed 5, low tm 2...</td>\n",
       "      <td>65.0</td>\n",
       "      <td>41.005307</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.113248</td>\n",
       "      <td>CHR1:5839-5859</td>\n",
       "      <td>6.191775</td>\n",
       "      <td>...</td>\n",
       "      <td>40.049105</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.455005</td>\n",
       "      <td>CHR1:5874-5894</td>\n",
       "      <td>13.838267</td>\n",
       "      <td>5.482251</td>\n",
       "      <td>GCTTCCAATGGAGCTTCTGC</td>\n",
       "      <td>59.544995</td>\n",
       "      <td>5581.0</td>\n",
       "      <td>5599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROMOSOME   END PRIMER_INTERNAL  \\\n",
       "21      CHR11   395       (195, 20)   \n",
       "20      CHR11   395       (195, 20)   \n",
       "3        CHR1  5894       (258, 20)   \n",
       "0        CHR1  5894       (258, 20)   \n",
       "2        CHR1  5894       (258, 20)   \n",
       "\n",
       "                              PRIMER_INTERNAL_EXPLAIN  \\\n",
       "21  considered 3185, GC content failed 3, low tm 2...   \n",
       "20  considered 3185, GC content failed 3, low tm 2...   \n",
       "3   considered 3005, GC content failed 5, low tm 2...   \n",
       "0   considered 3355, GC content failed 5, low tm 2...   \n",
       "2   considered 3005, GC content failed 5, low tm 2...   \n",
       "\n",
       "    PRIMER_INTERNAL_GC_PERCENT  PRIMER_INTERNAL_HAIRPIN_TH  \\\n",
       "21                        70.0                   41.637802   \n",
       "20                        70.0                   41.637802   \n",
       "3                         65.0                   41.005307   \n",
       "0                         65.0                   41.005307   \n",
       "2                         65.0                   41.005307   \n",
       "\n",
       "    PRIMER_INTERNAL_NUM_RETURNED  PRIMER_INTERNAL_PENALTY  \\\n",
       "21                           2.0                 0.033108   \n",
       "20                           2.0                 0.033108   \n",
       "3                            2.0                 0.113248   \n",
       "0                            2.0                 0.113248   \n",
       "2                            2.0                 0.113248   \n",
       "\n",
       "   PRIMER_INTERNAL_REGION  PRIMER_INTERNAL_SELF_ANY_TH  ...    \\\n",
       "21          CHR11:287-307                     0.000000  ...     \n",
       "20          CHR11:287-307                     0.000000  ...     \n",
       "3          CHR1:5839-5859                     6.191775  ...     \n",
       "0          CHR1:5839-5859                     6.191775  ...     \n",
       "2          CHR1:5839-5859                     6.191775  ...     \n",
       "\n",
       "    PRIMER_RIGHT_HAIRPIN_TH PRIMER_RIGHT_NUM_RETURNED  PRIMER_RIGHT_PENALTY  \\\n",
       "21                 0.000000                       2.0              5.799681   \n",
       "20                 0.000000                       2.0              5.799681   \n",
       "3                 40.049105                       2.0              0.455005   \n",
       "0                 40.049105                       2.0              0.455005   \n",
       "2                 40.049105                       2.0              0.455005   \n",
       "\n",
       "   PRIMER_RIGHT_REGION  PRIMER_RIGHT_SELF_ANY_TH PRIMER_RIGHT_SELF_END_TH  \\\n",
       "21       CHR11:372-395                  0.000000                 0.000000   \n",
       "20       CHR11:372-395                  0.000000                 0.000000   \n",
       "3       CHR1:5874-5894                 13.838267                 5.482251   \n",
       "0       CHR1:5874-5894                 13.838267                 5.482251   \n",
       "2       CHR1:5874-5894                 13.838267                 5.482251   \n",
       "\n",
       "      PRIMER_RIGHT_SEQUENCE  PRIMER_RIGHT_TM  REF_OFFSET  START  \n",
       "21  ACTCTACTACTCGAAGTTAGCAC        57.200319        92.0    128  \n",
       "20  ACTCTACTACTCGAAGTTAGCAC        57.200319        92.0    127  \n",
       "3      GCTTCCAATGGAGCTTCTGC        59.544995      5581.0   5599  \n",
       "0      GCTTCCAATGGAGCTTCTGC        59.544995      5581.0   5634  \n",
       "2      GCTTCCAATGGAGCTTCTGC        59.544995      5581.0   5599  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.amplicons_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>END</th>\n",
       "      <th>PRIMER</th>\n",
       "      <th>PRIMER_END_STABILITY</th>\n",
       "      <th>PRIMER_GC_PERCENT</th>\n",
       "      <th>PRIMER_HAIRPIN_TH</th>\n",
       "      <th>PRIMER_PENALTY</th>\n",
       "      <th>PRIMER_SELF_ANY_TH</th>\n",
       "      <th>PRIMER_SELF_END_TH</th>\n",
       "      <th>PRIMER_SEQUENCE</th>\n",
       "      <th>PRIMER_TM</th>\n",
       "      <th>REF_OFFSET</th>\n",
       "      <th>START</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5623</td>\n",
       "      <td>(18, 24)</td>\n",
       "      <td>3.66</td>\n",
       "      <td>37.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.649557</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TGAAACAAATTGACTTGGACACAG</td>\n",
       "      <td>58.350443</td>\n",
       "      <td>5581.0</td>\n",
       "      <td>5599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5622</td>\n",
       "      <td>(18, 23)</td>\n",
       "      <td>3.72</td>\n",
       "      <td>34.782609</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.651387</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TGAAACAAATTGACTTGGACACA</td>\n",
       "      <td>57.348613</td>\n",
       "      <td>5581.0</td>\n",
       "      <td>5599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5656</td>\n",
       "      <td>(53, 22)</td>\n",
       "      <td>4.75</td>\n",
       "      <td>40.909091</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.754965</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ACAGTCAATAGCAGAAAAGGCA</td>\n",
       "      <td>58.245035</td>\n",
       "      <td>5581.0</td>\n",
       "      <td>5634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5704</td>\n",
       "      <td>(29, 21)</td>\n",
       "      <td>3.71</td>\n",
       "      <td>52.380952</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.134189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGGATCGCTTGAGTAGAGGA</td>\n",
       "      <td>60.134189</td>\n",
       "      <td>5654.0</td>\n",
       "      <td>5683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHR1</td>\n",
       "      <td>5705</td>\n",
       "      <td>(30, 21)</td>\n",
       "      <td>3.85</td>\n",
       "      <td>57.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.002251</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CGGATCGCTTGAGTAGAGGAC</td>\n",
       "      <td>60.002251</td>\n",
       "      <td>5654.0</td>\n",
       "      <td>5684</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROMOSOME   END    PRIMER  PRIMER_END_STABILITY  PRIMER_GC_PERCENT  \\\n",
       "4        CHR1  5623  (18, 24)                  3.66          37.500000   \n",
       "6        CHR1  5622  (18, 23)                  3.72          34.782609   \n",
       "0        CHR1  5656  (53, 22)                  4.75          40.909091   \n",
       "14       CHR1  5704  (29, 21)                  3.71          52.380952   \n",
       "12       CHR1  5705  (30, 21)                  3.85          57.142857   \n",
       "\n",
       "    PRIMER_HAIRPIN_TH  PRIMER_PENALTY  PRIMER_SELF_ANY_TH  PRIMER_SELF_END_TH  \\\n",
       "4                 0.0        5.649557                 0.0                 0.0   \n",
       "6                 0.0        5.651387                 0.0                 0.0   \n",
       "0                 0.0        3.754965                 0.0                 0.0   \n",
       "14                0.0        1.134189                 0.0                 0.0   \n",
       "12                0.0        1.002251                 0.0                 0.0   \n",
       "\n",
       "             PRIMER_SEQUENCE  PRIMER_TM  REF_OFFSET  START  \n",
       "4   TGAAACAAATTGACTTGGACACAG  58.350443      5581.0   5599  \n",
       "6    TGAAACAAATTGACTTGGACACA  57.348613      5581.0   5599  \n",
       "0     ACAGTCAATAGCAGAAAAGGCA  58.245035      5581.0   5634  \n",
       "14     TCGGATCGCTTGAGTAGAGGA  60.134189      5654.0   5683  \n",
       "12     CGGATCGCTTGAGTAGAGGAC  60.002251      5654.0   5684  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.primers_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can analyse and filter these beds through d.primers.analyse or d.amplicons.analyse, etc..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "#     "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
